File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.kt
/*
주차 요금 계산 / 코틀린

출차 기록x 23:59간주
기본 시가 이하 -> 기본요금
기본 시간 초과 -> 기본요금 + 단위시간마다 단위요금

*/
class Solution {
    data class Park(
        var hour: Int,
        var minute: Int,
        var carNumber: String,
        var isIn: Boolean,
        var time : Int = 0,
        var fee : Int = 0,
    )

    fun splitRecord(record: String): Park {
        val list = record.split(' ')
        val hour = list[0].split(':')[0].toInt()
        val minute = list[0].split(':')[1].toInt()
        val car: String = list[1]
        val isIn: Boolean = list[2] == "IN"

        return Park(hour, minute, car, isIn)
    }

    private fun calculateFee(time : Int, fees : IntArray) : Int{
        if(time <= fees[0])
            return fees[1]
        else{
            val remain : Int = if((time - fees[0]) % fees[2] > 0) 1 else 0
            return fees[1] + ((time - fees[0]) / fees[2] + remain) * fees[3]
        }
    }

    fun solution(fees: IntArray, records: Array<String>): IntArray {
        var answer: IntArray = intArrayOf()
        val list = mutableListOf<Park>()


        records.forEachIndexed { index, s ->
            val record = splitRecord(s)
            val findCar = list.find {it.carNumber == record.carNumber}

            if(findCar == null){
                list.add(Park(record.hour, record.minute, record.carNumber, true, 0))
            }else {
                list.remove(findCar)
                if(record.isIn){
                    list.add(findCar.copy(record.hour, record.minute, record.carNumber, true, findCar.time))
                }else{
                    val diffTime = (record.hour - findCar.hour) * 60 + (record.minute - findCar.minute)
                    list.add(findCar.copy(record.hour, record.minute, record.carNumber, false, findCar.time + diffTime))
                }
            }
        }

        list.forEachIndexed { index, park ->
            if(list[index].isIn){
                val diffTime = (23 * 60 + 59) - (park.hour * 60 + park.minute)
                list[index].time += diffTime
                list[index].fee = calculateFee(list[index].time,fees)
            }else{
                list[index].fee = calculateFee(list[index].time, fees)
            }
        }

        return list.sortedBy { it.carNumber }.map { it.fee }.toIntArray()
    }
}

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.cpp
/*
주차 요금 계산

출차 기록 x -> 23:59 출차 간주
기본 시간 이하 -> 기본요금
기본시간 초과 -> 기본요금 + (초과분 단위시간 마다 단위요금 (올림 나머지 0아니면 무조건 올림))

fees / records 

fees
기본 시간 / 기본요금 / 단위시간 / 단위요금
records
시각 / 차량번호 / 내역

우선 시간은 일괄적 분으로 치환 -> 06 : 30은 390
*/
#include <string>
#include <vector>
#include <map>
#include <iostream>

using namespace std;

int timeToMinutes(string time);
bool checkIsIn(string s);
int calculateSumFee(const vector<int>& fees, int time);

vector<int> solution(vector<int> fees, vector<string> records) {
    vector<int> answer;
    
    //둘 다 key는 차 넘버
    map<string, int> m;
    map<string, int> sumMap;
    
    for(string s : records) {
        int time = timeToMinutes(s.substr(0,5));
        string carNumber = s.substr(6,4);
        bool isIn = checkIsIn(s.substr(11));
        
        if(isIn){
            m[carNumber] = time;
        }else {
            time -= m[carNumber];
            
            if(sumMap.find(carNumber) == sumMap.end()){
                sumMap.insert({carNumber, time});
            }else{
                sumMap[carNumber] += time;
            }
            m.erase(carNumber);
        }
    }
    
    for(auto it = m.begin(); it != m.end(); it++){
        string carNumber = it -> first;
        int inTime = it -> second;
        
        sumMap[carNumber] += (23 * 60 + 59) - inTime;
    }
    
    for(auto it = sumMap.begin(); it != sumMap.end(); it++){
        answer.push_back(calculateSumFee(fees, it->second));
    }
    
    
    return answer;
}

int timeToMinutes(string time) {
    int hour = stoi(time.substr(0,2));
    int minute = stoi(time.substr(3));
    
    return hour * 60  + minute;
}

bool checkIsIn(string s){
    return s == "IN";
}

int calculateSumFee(const vector<int>& fees, int time){
    if(time <= fees[0]){
        return fees[1];
    }else{
        time -= fees[0];
        return fees[1] + ((time / fees[2]) + (time % fees[2] > 0 ? 1 : 0)) * fees[3];
    }
}